src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Nuget/NugetRestorer.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.Threading;
+using NuGet.SolutionRestoreManager;
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Nuget
+{
+    using TIdentityDictionary = IImmutableDictionary<NamedIdentity, IComparable>;
+
+    internal class NugetRestorer : OnceInitializedOnceDisposedAsync
+    {
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private IVsSolutionRestoreService _solutionRestoreService;
+        private ImmutableHashSet<string> _watchedRules;
+        private IDisposable _evaluationSubscriptionLink;        
+
+        [ImportingConstructor]
+        public NugetRestorer(
+            IUnconfiguredProjectVsServices projectVsServices,
+            IVsSolutionRestoreService solutionRestoreService) 
+            : base(projectVsServices.ThreadingService.JoinableTaskContext)
+        {
+            Requires.NotNull(projectVsServices, nameof(projectVsServices));
+
+            _solutionRestoreService = solutionRestoreService;
+
+            _projectVsServices = projectVsServices;
+            _watchedRules = Empty.OrdinalIgnoreCaseStringSet
+                                 .Add(ConfigurationGeneral.SchemaName)
+                                 .Add(ProjectReference.SchemaName)
+                                 .Add(PackageReference.SchemaName);
+        }
+
+        public UnconfiguredProject Project
+        {
+            get
+            {
+                return _projectVsServices.Project;
+            }
+        }
+
+        [ProjectAutoLoad(startAfter: ProjectLoadCheckpoint.ProjectFactoryCompleted)]
+        [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+        internal async Task OnProjectFactoryCompletedAsync()
+        {
+            await InitializeCoreAsync(CancellationToken.None).ConfigureAwait(false);
+        }
+
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            ResetSubscriptions();
+
+            await TplExtensions.CompletedTask.ConfigureAwait(false);
+        }
+
+        protected override Task DisposeCoreAsync(bool initialized)
+        {
+            _evaluationSubscriptionLink?.Dispose();
+            return TplExtensions.CompletedTask;
+        }
+
+        private void ResetSubscriptions()
+        {
+            _evaluationSubscriptionLink?.Dispose();
+
+            var currentProjects = Project.LoadedConfiguredProjects;
+
+            if (currentProjects.Any())
```

Comment:
We don't want to listen to _all_ configured projects - just the ones that are considered active unique TFMs. 

For example, imagine I have the following configurations:

DBG|X86|NET45
DBG|X86|NET46
REL|X86|NET45
REL|X86|NET46

And my "active" traits are "DBG" and "X86" then you only want to listen to:

DBG|X86|NET45
DBG|X86|NET46

The other results are not interesting - and resulting you doing extra work and book keeping. You should with Manish, because we should probably introduce a service that provides that information as "active configured projects"